renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -131,8 +131,645 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
+                           uv.value.f.y, uv.value.f.z, uv.value.f.w);
+  else
+    StringFormat::snprintf(buf, 255, formats[texdim + texdimOffs - 1][texcoordType], uv.value.i.x,
+                           uv.value.i.y, uv.value.i.z, uv.value.i.w);
+
+  if(ddxType == 0)
+    StringFormat::snprintf(buf2, 255, formats[offsetDim + texdimOffs - 1][ddxType], ddxCalc.value.f.x,
+                           ddxCalc.value.f.y, ddxCalc.value.f.z, ddxCalc.value.f.w);
+  else
+    StringFormat::snprintf(buf2, 255, formats[offsetDim + texdimOffs - 1][ddxType], ddxCalc.value.i.x,
+                           ddxCalc.value.i.y, ddxCalc.value.i.z, ddxCalc.value.i.w);
+
+  if(ddyType == 0)
+    StringFormat::snprintf(buf3, 255, formats[offsetDim + texdimOffs - 1][ddyType], ddyCalc.value.f.x,
+                           ddyCalc.value.f.y, ddyCalc.value.f.z, ddyCalc.value.f.w);
+  else
+    StringFormat::snprintf(buf3, 255, formats[offsetDim + texdimOffs - 1][ddyType], ddyCalc.value.i.x,
+                           ddyCalc.value.i.y, ddyCalc.value.i.z, ddyCalc.value.i.w);
+
+  std::string texcoords = buf;
+  std::string ddx = buf2;
+  std::string ddy = buf3;
+
+  if(opcode == OPCODE_LD_MS)
+  {
+    StringFormat::snprintf(buf, 255, formats[0][1], multisampleIndex);
+  }
+
+  std::string sampleIdx = buf;
+  std::string offsets = "";
+
+  if(useOffsets)
+  {
+    if(offsetDim == 1)
+      StringFormat::snprintf(buf, 255, ", int(%d)", texelOffsets[0]);
+    else if(offsetDim == 2)
+      StringFormat::snprintf(buf, 255, ", int2(%d, %d)", texelOffsets[0], texelOffsets[1]);
+    else if(offsetDim == 3)
+      StringFormat::snprintf(buf, 255, ", int3(%d, %d, %d)", texelOffsets[0], texelOffsets[1],
+                             texelOffsets[2]);
+    // texdim == 4 is cube arrays, no offset supported
+
+    offsets = buf;
+  }
+
+  char elems[] = "xyzw";
+  std::string strSwizzle = ".";
+  for(int i = 0; i < 4; ++i)
+  {
+    strSwizzle += elems[swizzle[i]];
+  }
+
+  std::string strGatherChannel;
+  switch(gatherChannel)
+  {
+    case ShaderDebug::GatherChannel::Red: strGatherChannel = "Red"; break;
+    case ShaderDebug::GatherChannel::Green: strGatherChannel = "Green"; break;
+    case ShaderDebug::GatherChannel::Blue: strGatherChannel = "Blue"; break;
+    case ShaderDebug::GatherChannel::Alpha: strGatherChannel = "Alpha"; break;
+  }
+
+  std::string vsProgram = "float4 main(uint id : SV_VertexID) : SV_Position {\n";
+  vsProgram += "return float4((id == 2) ? 3.0f : -1.0f, (id == 0) ? -3.0f : 1.0f, 0.5, 1.0);\n";
+  vsProgram += "}";
+
+  std::string sampleProgram;
+
+  UINT texSlot = resourceData.slot;
+  UINT sampSlot = samplerData.slot;
+
+  if(opcode == OPCODE_SAMPLE || opcode == OPCODE_SAMPLE_B || opcode == OPCODE_SAMPLE_D)
+  {
+    sampleProgram = textureDecl + " : register(t" + std::to_string(texSlot) + ");\n" + samplerDecl +
+                    " : register(s" + std::to_string(sampSlot) + ");\n\n";
```

Comment:
Prefer not using `std::to_string`, if you need an int-to-string then `ToStr` will do the trick, or for more complex formatting like this you might want to use `StringFormat::Fmt`.

(I know the D3D11 and this code uses `StringFormat::snprintf` which is fine for now, but that's super old and predates the existance of `StringFormat::Fmt`. I plan on removing the `snprintf` variant so using `Fmt` is the way to go for formatting in purely new code).

Reply:
Good to know. Migrated to StringFormat::Fmt for the places that were using std::to_string, and removed a few snprintf calls as well.